quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/SyncFunction.java

```diff
@@ -112,32 +109,38 @@ public void setColumnsToMigrate(Set<String> columnsToMigrate) {
 		this.insertExpressions = ImmutableMap.copyOf(expressions);
 		this.updateExpressions = ImmutableMap.copyOf(insertExpressions);
 
-		this.updateIdentitiesForInserts = ImmutableMap.copyOf(targetTable.getIdentityColumns().stream()
+		this.updateIdentitiesForInserts = ImmutableMap.copyOf((Map<? extends String, ? extends String>)targetTable.getColumns().stream()
+				.filter(column -> reverseLookup(mapping, column.getName()).isPresent())
 				.collect(Collectors.toMap(column -> "\"" + column.getName() + "\"",
-						column -> "NEW.\"" + reverseLookup(mapping, column.getName()) + "\"",
+						column -> "NEW.\"" + reverseLookup(mapping, column.getName()).get() + "\"",
 						(u, v) -> { throw new IllegalStateException(String.format("Duplicate key %s", u)); },
 						Maps::newLinkedHashMap)));
 
-		this.updateIdentities = ImmutableMap.copyOf(targetTable.getIdentityColumns().stream()
+		this.updateIdentities = ImmutableMap.copyOf((Map<? extends String, ? extends String>)targetTable.getIdentityColumns().stream()
+				.filter(column -> reverseLookup(mapping, column.getName()).isPresent())
 				.collect(Collectors.toMap(column -> "\"" + column.getName() + "\"",
-						column -> "OLD.\"" + reverseLookup(mapping, column.getName()) + "\"",
+						column -> "OLD.\"" + reverseLookup(mapping, column.getName()).get() + "\"",
 						(u, v) -> { throw new IllegalStateException(String.format("Duplicate key %s", u)); },
 						Maps::newLinkedHashMap)));
 	}
 
-	private String reverseLookup(Map<String, String> mapping, String value) {
+	private Optional<String> reverseLookup(Map<String, String> mapping, String value) {
 		return mapping.entrySet().stream()
 				.filter(entry -> entry.getValue().equals(value))
 				.findFirst()
-				.map(Entry::getKey)
-				.get();
+				.map(Entry::getKey);
 	}
 
 	public QueryBuilder createFunctionStatement() {
 		return new QueryBuilder()
```

Comment:
> I also redesigned the cascading trigger functionality. I tested more than 2 active versions and it did not update version 3 (or 4 for that matter) of a table as the trigger triggering version 2 could not trigger the one going to version 3. (V1 <-> V2 <-> V3 <-> V4 ...)

The intended way for the trigger to function is to only update one table. We want to avoid cascading updates as you described because they're pretty complex and hard to follow / debug.

Instead what should happen with >2 active versions is that you form a fully connected graph using triggers:

```
  V1 <-------> V2 <-------> V3
   ^                        ^
   |                        |
    ------------------------
```

So for a give table active in 3 versions, you'd create sync functions that
- Change table X in V2 when table X in V1 is changed.
- Change table X in V3 when table X in V1 is changed.
- Change table X in V1 when table X in V2 is changed.
- Change table X in V3 when table X in V2 is changed.
- Change table X in V1 when table X in V3 is changed.
- Change table X in V2 when table X in V3 is changed.

> This also prevents the endless loop for which pg_trigger_depth() = 0 was used.

`pg_trigger_depth() = 0` actually prevents triggering other triggers. Not having it will indeed cause endless loops.

Reply:
I think it will just be easier to only create the triggers between subsequent versions. If you have to take into account all the additional changes between, for example, V1 and V4 when you could just let the other V2 and V3 triggers deal with their own (smaller) changes, this would decrease the complexity a lot. The way I now implemented it is that is can only propagate in one way and never go back and visit the previous one.

Additionaly, having pg_trigger_depth() = 0 would just completely disable any manually made triggers (they are not supported yet, but if they were this would not work).

Reply:
> I think it will just be easier to only create the triggers between subsequent versions. If you have to take into account all the additional changes between, for example, V1 and V4 when you could just let the other V2 and V3 triggers deal with their own (smaller) changes, this would decrease the complexity a lot. The way I now implemented it is that is can only propagate in one way and never go back and visit the previous one.

See my previous point why this doesn't work for the more complex migrations.

> Additionaly, having pg_trigger_depth() = 0 would just completely disable any manually made triggers (they are not supported yet, but if they were this would not work).

That's not correct. We're using an "after" trigger to sync the data, which is a trigger that runs after any "before" triggers that want to modify the data before persisting it. You can still safely use other "manually made" triggers. Here's an example with SQL so you can reproduce it locally:

```sql
-- Create base tables A and B
CREATE TABLE a (id bigserial, name text, last_modified timestamptz, primary key (id));
CREATE TABLE b (id bigserial, name text, last_modified timestamptz, primary key (id));


-- Create data triggers for A and B, which update the last_modified column anytime something is changed.
CREATE OR REPLACE FUNCTION set_last_modified() RETURNS TRIGGER AS $$
	DECLARE result record;
	BEGIN
		NEW.last_modified = statement_timestamp();
		RETURN NEW;
	END;
$$ LANGUAGE 'plpgsql';

CREATE TRIGGER t3
BEFORE INSERT OR UPDATE
ON a
FOR EACH ROW
EXECUTE PROCEDURE set_last_modified();

CREATE TRIGGER t4
BEFORE INSERT OR UPDATE
ON b
FOR EACH ROW
EXECUTE PROCEDURE set_last_modified();


-- Create triggers and update functions between tables A and B
CREATE OR REPLACE FUNCTION f1()
RETURNS TRIGGER AS $$
BEGIN
  IF TG_OP = 'INSERT' THEN
    INSERT INTO b (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
  ELSIF TG_OP = 'UPDATE' THEN
    LOOP
      UPDATE b
        SET id = NEW.id, name = NEW.name, last_modified = NEW.last_modified
        WHERE id = OLD.id;
      IF found THEN EXIT; END IF;
      BEGIN
        INSERT INTO b (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
      EXIT;
      EXCEPTION WHEN unique_violation THEN END;
	END LOOP;
  ELSIF TG_OP = 'DELETE' THEN
    DELETE FROM b
      WHERE id = OLD.id;
  END IF;
  RETURN NEW;
END;
$$ LANGUAGE 'plpgsql';


CREATE TRIGGER t1
AFTER INSERT OR UPDATE OR DELETE
ON a
FOR EACH ROW
WHEN (pg_trigger_depth() = 0)
EXECUTE PROCEDURE f1();


CREATE OR REPLACE FUNCTION f2()
RETURNS TRIGGER AS $$
BEGIN
  IF TG_OP = 'INSERT' THEN
    INSERT INTO a (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
  ELSIF TG_OP = 'UPDATE' THEN
    LOOP
      UPDATE a
        SET id = NEW.id, name = NEW.name, last_modified = NEW.last_modified
        WHERE id = OLD.id;
      IF found THEN EXIT; END IF;
      BEGIN
        INSERT INTO a (id, name, last_modified) VALUES (NEW.id, NEW.name, NEW.last_modified);
      EXIT;
      EXCEPTION WHEN unique_violation THEN END;
	END LOOP;
  ELSIF TG_OP = 'DELETE' THEN
    DELETE FROM a
      WHERE id = OLD.id;
  END IF;
  RETURN NEW;
END;
$$ LANGUAGE 'plpgsql';


CREATE TRIGGER t2
AFTER INSERT OR UPDATE OR DELETE
ON b
FOR EACH ROW
WHEN (pg_trigger_depth() = 0)
EXECUTE PROCEDURE f2();


-- Example showcasing "manual" triggers. Both records printed will show the updated "last_modified" column.
INSERT INTO a (name) VALUES ('Michael') RETURNING *;
SELECT * FROM b;
```